File: app/src/main/java/com/example/ogl4jo3/accounting/setting/categorymanagement/income/IncomeCategoryAdapter.java
package com.example.ogl4jo3.accounting.setting.categorymanagement.income;

import android.app.FragmentManager;
import android.content.Context;
import android.database.sqlite.SQLiteDatabase;
import android.util.Log;
import android.widget.Toast;

import com.example.ogl4jo3.accounting.R;
import com.example.ogl4jo3.accounting.setting.categorymanagement.Category;
import com.example.ogl4jo3.accounting.setting.categorymanagement.CategoryAdapter;
import com.example.ogl4jo3.accounting.setting.categorymanagement.CategoryDAO;
import com.example.ogl4jo3.utility.database.MyDBHelper;
import com.google.gson.Gson;

import java.util.Collections;
import java.util.List;

import static android.content.ContentValues.TAG;

/**
 * 收入類別itemAdapter
 * Created by ogl4jo3 on 2017/7/14.
 */

public class IncomeCategoryAdapter extends CategoryAdapter {

	private Context mContext;
	private FragmentManager fragmentManager;
	private List<Category> categoryList;

	public IncomeCategoryAdapter(Context context, FragmentManager fragmentManager,
	                             List<Category> categoryList) {
		super(context, fragmentManager, categoryList);
		this.mContext = context;
		this.fragmentManager = fragmentManager;
		this.categoryList = categoryList;
	}

	@Override
	public void onItemClick(int position) {
		String categoryJson = (new Gson()).toJson(categoryList.get(position));
		IncomeCategoryNewEditFragment incomeCategoryNewEditFragment =
				IncomeCategoryNewEditFragment.newInstance("edit", categoryJson);
		fragmentManager.beginTransaction()
				.replace(R.id.layout_main_content, incomeCategoryNewEditFragment, null)
				.addToBackStack(null).commit();

		//Toast.makeText(mContext, "Click on " + categoryList.get(position).getName(),
		//		Toast.LENGTH_SHORT).show();
	}

	@Override
	public void onItemMove(int fromPosition, int toPosition) {
		//更新排列順序
		SQLiteDatabase db = MyDBHelper.getDatabase(mContext);
		new CategoryDAO(db)
				.updateIncomeOrderNum(categoryList.get(fromPosition), categoryList.get(toPosition));

		//即時更新資料
		int fromOrderNum = categoryList.get(fromPosition).getOrderNum();
		int toOrderNum = categoryList.get(toPosition).getOrderNum();
		categoryList.get(fromPosition).setOrderNum(toOrderNum);
		categoryList.get(toPosition).setOrderNum(fromOrderNum);

		Log.d(TAG, "onItemMove: from:" + fromPosition + ", to:" + toPosition);

		Collections.swap(categoryList, fromPosition, toPosition);
		notifyItemMoved(fromPosition, toPosition);

	}

	@Override
	public void onItemDismiss(int position) {
		//刪除
		SQLiteDatabase db = MyDBHelper.getDatabase(mContext);
		new CategoryDAO(db).deleteIncomeData(categoryList.get(position));

		Toast.makeText(mContext, mContext.getResources()
						.getString(R.string.tv_category_deleted, categoryList.get(position).getName()),
				Toast.LENGTH_SHORT).show();
		categoryList.remove(position);
		notifyItemRemoved(position);

	}
}

File: app/src/main/kotlin/com/ogl4jo3/accounting/ui/categoryMgmt/IncomeCategoryAdapter.kt
package com.ogl4jo3.accounting.ui.categoryMgmt

import android.content.Context
import android.widget.Toast
import androidx.fragment.app.FragmentManager
import androidx.navigation.Navigation
import com.ogl4jo3.accounting.R
import com.ogl4jo3.accounting.data.CategoryType
import com.ogl4jo3.accounting.setting.categorymanagement.Category
import com.ogl4jo3.accounting.setting.categorymanagement.CategoryDAO
import com.ogl4jo3.accounting.utils.database.MyDBHelper
import timber.log.Timber
import java.util.Collections

/**
 * 收入類別itemAdapter
 * Created by ogl4jo3 on 2017/7/14.
 */
class IncomeCategoryAdapter(
    private val mContext: Context, fragmentManager: FragmentManager,
    private val categoryList: List<Category>
) : CategoryAdapter(mContext, fragmentManager, categoryList) {
    override fun onItemClick(holder: ViewHolder) {
        Navigation.findNavController(holder.itemView).navigate(
            IncomeCategoryMgmtFragmentDirections.actionIncomeCategoryMgmtFragmentToIncomeCategoryEditFragment(
                //TODO: workaround , for test
                category = com.ogl4jo3.accounting.data.Category(
                    id = "a1110637-5ad5-449b-a09e-6886228aa3c9",
                    orderNumber = 0,
                    name = "test",
                    iconResName = mContext.resources.getResourceEntryName(R.drawable.ic_category_afternoon_tea),
                    categoryType = CategoryType.Expense
                )
//                title = mContext.getString(R.string.title_income_category_edit),
//                categoryJsonStr = Gson().toJson(categoryList[holder.adapterPosition])
            )
        )
        //Toast.makeText(mContext, "Click on " + categoryList.get(position).getName(),
        //		Toast.LENGTH_SHORT).show();
    }

    override fun onItemMove(fromPosition: Int, toPosition: Int) {
        //更新排列順序
        val db = MyDBHelper.getDatabase(mContext)
        CategoryDAO(db)
            .updateIncomeOrderNum(categoryList[fromPosition], categoryList[toPosition])

        //即時更新資料
        val fromOrderNum = categoryList[fromPosition].orderNum
        val toOrderNum = categoryList[toPosition].orderNum
        categoryList[fromPosition].orderNum = toOrderNum
        categoryList[toPosition].orderNum = fromOrderNum
        Timber.d("onItemMove: from:$fromPosition, to:$toPosition")
        Collections.swap(categoryList, fromPosition, toPosition)
        notifyItemMoved(fromPosition, toPosition)
    }

    override fun onItemDismiss(position: Int) {
        //刪除
        val db = MyDBHelper.getDatabase(mContext)
        CategoryDAO(db).deleteIncomeData(categoryList[position])
        Toast.makeText(
            mContext, mContext.getString(
                R.string.msg_category_deleted,
                categoryList[position].name
            ), Toast.LENGTH_SHORT
        ).show()
        categoryList.toMutableList().removeAt(position)//TODO: workaround, check remove method
        notifyItemRemoved(position)
    }
}